### PR TITLE
(feat:grid) update selected rows when filters are applied

### PR DESCRIPTION
**Pull request by kdemirov**

### CODE REVIEW

File: projects/data-grid/src/lib/store/data-grid.ts [hunk 1]
```diff
@@ -379,7 +426,7 @@ const rowIndexesAndPaginationReducer = createReducer(initialGridState, on(
   GridActions.changePageNumber,
   GridActions.deleteRow,
   GridActions.addRow,
-  recalculateRowIndexesAndPagination
+  R.compose(recalculateSelectedRowsIndexes, recalculateRowIndexesAndPagination)
```
Comment (velinovskav) [inline projects/data-grid/src/lib/store/data-grid.ts]: It's better to have an `updateGridRowsHandler` that will recalculate row indexes, selected rows and pagination.

File: projects/data-grid/src/lib/store/data-grid.ts [hunk 2]
```diff
@@ -286,6 +313,26 @@ const addRow = (state: GridState, action: AddRowPayload<any>): GridState => {
...
+
+const recalculateSelectedRowsIndexes = (state: GridState): GridState => R.mergeRight(state,{
+    selectedRowsIndexes: R.cond([
+      [R.propEq('allPagesSelected', true), getRowDataIndexes],
+      [R.propEq('currentPageSelected', true), calculateSelectedRowsIfCurrentPageSelected],
+      [R.T, alignSelectedRowsIndexesWithRowDataIndexes]
+    ])(state)
+  });
+
 const recalculateRowIndexesAndPagination = (state: GridState): any => {
```
Comment (velinovskav) [inline projects/data-grid/src/lib/store/data-grid.ts]: we should make sure that when row indexes and pagination is recalculated, the selected rows are updated too

File: projects/data-grid/src/lib/store/data-grid.ts [hunk 3]
```diff
@@ -286,6 +313,26 @@ const addRow = (state: GridState, action: AddRowPayload<any>): GridState => {
...
+  const filteredSelectedRows = alignSelectedRowsIndexesWithRowDataIndexes(state);
+
+  return currentPageContainsSelectedRows(state)
+  ? getCurrentPageIndexes(state) : R.uniq(R.concat(getCurrentPageIndexes(state), filteredSelectedRows));
+};
+
+const recalculateSelectedRowsIndexes = (state: GridState): GridState => R.mergeRight(state,{
+    selectedRowsIndexes: R.cond([
+      [R.propEq('allPagesSelected', true), getRowDataIndexes],
+      [R.propEq('currentPageSelected', true), calculateSelectedRowsIfCurrentPageSelected],
```
Comment (velinovskav) [inline projects/data-grid/src/lib/store/data-grid.ts]: Shouldn't we name it something like `getRowIndexesOfTheCurrentPage`?

File: projects/data-grid/src/lib/store/data-grid.ts [hunk 4]
```diff
@@ -286,6 +313,26 @@ const addRow = (state: GridState, action: AddRowPayload<any>): GridState => {
...
+      [R.T, alignSelectedRowsIndexesWithRowDataIndexes]
```
Comment (velinovskav) [inline projects/data-grid/src/lib/store/data-grid.ts]: `getSelectedRowIndexesIncludedInVisibleRows` or just do filtering here

File: projects/data-grid/src/lib/store/data-grid.ts [hunk 5]
```diff
@@ -86,6 +94,13 @@ export const getPagination: GetPagination = R.propOr(initialPagination, 'paginat
 type GetChildren = (state: GridState) => string[];
 export const getChildren: GetChildren = R.propOr([], 'children');
 
+type GetCurrentPageIndexes = (state: GridState) => number[];
```
Comment (velinovskav) [inline projects/data-grid/src/lib/store/data-grid.ts]:
```suggestion
type GetCurrentPageRowIndexes = (state: GridState) => number[];
```

File: projects/data-grid/src/lib/store/data-grid.ts [hunk 6]
```diff
@@ -86,6 +94,13 @@ export const getPagination: GetPagination = R.propOr(initialPagination, 'paginat
...
+const getCurrentPageIndexes: GetCurrentPageIndexes = (state: GridState) => {
```
Comment (velinovskav) [inline projects/data-grid/src/lib/store/data-grid.ts]:
```suggestion
const getCurrentPageRowIndexes: GetCurrentPageRowIndexes = (state: GridState) => {
```

File: projects/data-grid/src/lib/store/data-grid.ts [hunk 7]
```diff
@@ -86,6 +94,13 @@ export const getPagination: GetPagination = R.propOr(initialPagination, 'paginat
...
+  const {rowDataIndexes, pagination} = state;
+  const {currentPage, paginationPageSize} = pagination;
+  return  getPagedData(rowDataIndexes, currentPage, paginationPageSize);
```
Comment (velinovskav) [inline projects/data-grid/src/lib/store/data-grid.ts]:
```suggestion
  return getPagedData(rowDataIndexes, currentPage, paginationPageSize);
```

File: projects/data-grid/src/lib/store/data-grid.ts [hunk 8]
```diff
@@ -154,19 +169,43 @@ const filterGridHandler = (state: GridState, {columnId, option, value}: FilterGr
...
+};
+
+const isCurrentPageSelected = (state: GridState) => R.cond([
+    [currentPageEqualsSelectedRows, R.always(true)],
+    [hasCustomSelection, currentPageContainsSelectedRows],
+    [R.T, R.always(false)]
+  ])(state);
+
 const changePageSizeHandler = (state: GridState, {pageSize}): GridState => R.mergeRight(state, {
   pagination: {...state.pagination, paginationPageSize: pageSize}
```
Comment (velinovskav) [inline projects/data-grid/src/lib/store/data-grid.ts]: here we should update `currentPageSelected ` and `allSelected`

File: projects/data-grid/src/lib/store/data-grid.ts [hunk 9]
```diff
@@ -154,19 +169,43 @@ const filterGridHandler = (state: GridState, {columnId, option, value}: FilterGr
...
 });
 
+const handleCurrentPage = (state: GridState) => R.mergeRight(state, {
+   currentPageSelected: !getAllPagesSelected(state) ? isCurrentPageSelected(state) : false,
```
Comment (velinovskav) [inline projects/data-grid/src/lib/store/data-grid.ts]: `currentPageSelected` should be `true` when `allPagesSelected` is `false`, and the length of the selected row indexes is equal to the length of the current row indexes, and the data in both lists is the same.

File: projects/data-grid/src/lib/store/data-grid.ts [hunk 10]
```diff
@@ -154,19 +169,43 @@ const filterGridHandler = (state: GridState, {columnId, option, value}: FilterGr
...
+   allSelected: currentPageContainsSelectedRows(state)
```
Comment (velinovskav) [inline projects/data-grid/src/lib/store/data-grid.ts]: shouldn't we calculate `allSelected` based on the following rule if `allPagesSelected` is true then the data in `rowDataIndexes` and `selectedRowsIndexes` should be the same, else the data in `selectedRowsIndexes` and current row indexes should be the same.

File: projects/data-grid/src/lib/store/data-grid.spec.ts [hunk 11]
```diff
@@ -646,4 +648,162 @@ describe('Data Grid reducer', () => {
     expect(result.data[7].id).toEqual(45);
   });
 
+  it('should recalculate selected row indexes when update filter is dispatched and current page is selected', () => {
```
Comment (velinovskav) [inline projects/data-grid/src/lib/store/data-grid.spec.ts]: maybe it would be smart to split tests in different files

**Review by velinovskav — CHANGES_REQUESTED** [review]

File: projects/data-grid/src/lib/store/data-grid.ts (hunk 2 above)
Comment (kdemirov) [reply projects/data-grid/src/lib/store/data-grid.ts]: I think that this will be handled with the` updateGridRowsHandler` or i am missing something ?

Comment (velinovskav) [reply projects/data-grid/src/lib/store/data-grid.ts]: if we put all the logic there yes

File: projects/data-grid/src/lib/store/data-grid.ts (hunk 10 above)
Comment (kdemirov) [reply projects/data-grid/src/lib/store/data-grid.ts]: I have a question here, for example we select all the elements from the first page then we navigate to second page and select only one item, we go back to the first page, so now the `selectedRows` are the indexes from the first page plus the selected item from the second page should `allSelected` remain true or should it be false.

Comment (velinovskav) [reply projects/data-grid/src/lib/store/data-grid.ts]: IMO once we select something different than only items in one page, the value should be set to false

**Review by velinovskav — APPROVED** [review]

**Review by trajchen — APPROVED** [review]

**Review by anaivanovska — APPROVED** [review]

**Pull request merged**